Updates Documentation for Sendgrid (#36)
triggermesh/docs: docs/targets/sendgrid.md

# Event Target for Sendgrid

This event target receives [CloudEvents][ce] and utilizes [Sendgrid][sg] to enable the creation and delivery of Email messages via event-data and event-occurrence, respectively.

## Prerequisites

* Register a [Sendgrid account][sgSU]
* Retrieve a [Sendgrid API token][api]

## Deploying an Instance of the Target

Open the Bridge creation screen and add a target of type `Sendgrid`.

![Adding a Sendgrid target](../images/sendgrid-target/create-bridge-1.png)

> In the Target creation form, give a name to the event Target and add the following information:

Reviewer: How about `give a name to` with `provide a name for`?

Reviewer: May need to evaluate the comma usage as I believe there should be one before `and`.

* **Default sender name**: Assign a default sender name. (optional)
* **Default sender email**: Assign a default sender email. (optional)
* **Default recipient name**:  Assign a default recipient's  name. (optional)

> * **Default recipient email**: Assign a default recipient's email. (optional)

Reviewer 2: _Assign_... to what? I have never used that target and am puzzled after reading this. From a user perspective it would be better for formulate these as

> Sender name to assign to XYZ if the received event does not already contain that information.

or something else along those lines, as long as it tells what the field is used for.

> * **API Secret**: Reference to a [TriggerMesh secret][tm-secret] containing a [API token][api] for authenticating requests

Reviewer: Should be `an` before API.

> ![Sendgridtarget form](../images/sendgrid-target/create-bridge-2.png)

Reviewer 2: ```suggestion
![SendgridTarget form](../images/sendgrid-target/create-bridge-2.png)
```

After clicking the `Save` button, the console will self-navigate to the Bridge editor. Proceed by adding the remaining components to the Bridge.

> After submitting the bridge, and allowing some configuration time, a green check mark on the main _Bridges_ page indicates that the bridge with an Elasticsearch event Target was successfully created.

Reviewer: We're using Elasticsearch now?

Reviewer 3: sorry lol. copy pasta :)

> After submitting the bridge, and allowing some configuration time, a green check mark on the main _Bridges_ page indicates that the bridge with a Sendgrid event Target was successfully created.

Reviewer 2: ```suggestion
After submitting the bridge, and allowing some configuration time, a green check mark on the main _Bridges_ page indicates that the bridge with a SendGrid event Target was successfully created.
```

![Bridge status](../images/bridge-status-green.png)

For more information about using SendGrid, please refer to the [SendGrid documentation][docs].

### Example

The SendGrid event Target can consume events of any type.

> This Target accepts a [JSON][ce-jsonformat] payload with the following properties:

Reviewer 2: This is in direct contradiction with

> The SendGrid event Target can consume events of any type.

above, so you can delete the sentence I quoted.

If we don't have a notion of "Type" in this target yet, it's better to just document the JSON format instead of saying the target accepts "any type".

Reviewer 3: but.. the sendgrid target **does** accept and will consume events of any type. Maybe I am just unclear on something. ill ping you :)

Reviewer 4: I think what Antoine is saying is that maybe the CE type can be anything but the JSON payload cannot be anything since you have specific keys.

Reviewer 4: looks like this is now corrected. so make your PR something real and not a DRAFT and then merge :)

Reviewer 3: I did speak with @antoineco on the matter and I have made updates after our conversation. If it is not resolved as-is, we are very close.

| Name  |  Type |  Comment | Required
|---|---|---|---|

> | **FromName**| string  |Sender's name. |false |
> | **FromEmail**| string  |Sender's email. | false |
> |  **ToEmail** |  string |Recipient's email.| false |
> |  **ToName** |  string |Recipient's name.| false |
> |**Message**|string| Contents of the message body |false |

Reviewer 2: This needs some formatting.
```suggestion
| **FromName** | string | Sender's name |false |
| **FromEmail** | string | Sender's email | false |
| **ToName** | string | Recipient's name | false |
| **ToEmail** | string | Recipient's email | false |
| **Message** | string | Contents of the message body | false |
```

> If the **Message** key value is omited within an event, the body of the email will contain the entire cloud event

Reviewer 2: Typo, formatting, missing period. 
```suggestion
If the **Message** key value is omitted within an event, the body of the email will contain the entire CloudEvent.
```

* **Default recipient name**: Define a default name to be assigned in the `To:` section of the email to be created, if the received event does not contain a **FromEmail** property.
* **Default recipient email**: Define a default 'email address' to be assigned in the `To:` section of the email to be created, if the received event does not contain a **ToEmail** property.
* **API Secret**: Reference to a [TriggerMesh secret][tm-secret] containing an [API token][api] for authenticating requests

![SendgridTarget form](../images/sendgrid-target/create-bridge-2.png)

After clicking the `Save` button, the console will self-navigate to the Bridge editor. Proceed by adding the remaining components to the Bridge.

After submitting the bridge, and allowing some configuration time, a green check mark on the main _Bridges_ page indicates that the bridge with a SendGrid event Target was successfully created.

![Bridge status](../images/bridge-status-green.png)

For more information about using SendGrid, please refer to the [SendGrid documentation][docs].

> ### Event Type

Reviewer 2: ```suggestion
### Event Types
```

> The SendGrid event target can consume events of any type[^s]

Reviewer 2: This funny character at the end does not render, check the rendered page.

Reviewer 3: that is odd.. it renders in my markdown editor but i guess GitHub doesn't like it for some reason.

Reviewer 3: oh.. looks like it is valid markdown but GitHub has an issue with footnotes. Ref-> https://github.com/github/markup/issues/498

Reviewer 3: fixing now :)

The SendGrid event Target accepts a [JSON][ce-jsonformat] payload with the following properties that will overwrite their respective `spec` parameters.[^s]

| Name  |  Type |  Comment | Required
|---|---|---|---|
| **FromName** | string | Sender's name |false |
| **FromEmail** | string | Sender's email | false |
| **ToName** | string | Recipient's name | false |
| **ToEmail** | string | Recipient's email | false |
| **Message** | string | Contents of the message body | false |

If the **Message** key value is omitted within an event, the body of the email will contain the entire CloudEvent.

### Example

The Target uses the `Event Source` to assign the `subject` of new emails in the following format:

```go

> "Notification from " + string(event.Source()) + " Knative Event"

Reviewer 3: Would @cab105  or @antoineco (or anyone) mind providing me with a better "front facing" format to display how the subject is formatted/propagated

Reviewer: Just using ``` without the language specifier will work fine.

Reviewer 3: While re-phrasing my question I came up with a way to write this that, i hope, is acceptable. 

Ref -> https://github.com/triggermesh/docs/blob/d8b6f0997c3776d5bc6539f267c901d1ced5d4e9/docs/targets/sendgrid.md#example

```

The entire cloud event is passed into the email `body` by default.

When a **Message** property is present in the payload, its associated value is used to propagate the `body` of the email.

[sgSU]:https://signup.sendgrid.com/
[sg]:https://sendgrid.com/
[api]:https://sendgrid.com/docs/ui/account-and-settings/api-keys/

[ce]: https://cloudevents.io/
[ce-jsonformat]: https://github.com/cloudevents/spec/blob/v1.0/json-format.md
[tm-secret]:https://docs.triggermesh.io/guides/secrets/
[docs]: https://sendgrid.com/docs/

> [^s]: *Note*: If there is not a default value specified at the spec for all of the available fields, an event *MUST* contain all of the following, save **Message**, or the Target **will** **fail**

Reviewer 2: This target works a bit strangely compared to other targets due to these two modes of operation. It makes it tricky to explain how things work and leaks in multiple places, like the description of fields in `Deploying an Instance of the Target`, which are only relevant for one of the two modes.

This is not a good place to discuss the design of the target, so the only suggestion that comes to my mind is to add a section after `Prerequisites` that explain the two modes in detail, and reference that section from `Event Types` section to avoid turning the description of event types into a dumpster. It also gives more immediate context to the reader and makes the rest of the document clearer.

Open the Bridge creation screen and add a target of type `SendGrid`.

![Adding a SendGrid target](../images/sendgrid-target/create-bridge-1.png)

In the Target creation form, provide a name for the event Target, and add the following information:

* **Default sender name**: Define a default 'name' to be assigned in the `From:` section of the email to be created, if the received event does not contain a **FromName** property.
* **Default sender email**: Define a default email address to be assigned in the `From:` section of the email to be created, if the received event does not contain a **FromEmail** property.
* **Default recipient name**: Define a default name to be assigned in the `To:` section of the email to be created, if the received event does not contain a **FromEmail** property.
* **Default recipient email**: Define a default 'email address' to be assigned in the `To:` section of the email to be created, if the received event does not contain a **ToEmail** property.
* **API Secret**: Reference to a [TriggerMesh secret][tm-secret] containing an [API token][api] for authenticating requests

![SendgridTarget form](../images/sendgrid-target/create-bridge-2.png)

> After selecting the `Save` button, the console will self-navigate to the Bridge editor. Proceed by adding the remaining components to the Bridge.

Reviewer: `clicking` was fine in this context.
